# **Competencia 1 - CC6205 Natural Language Processing 📚**

**Integrantes:** Nicolas Lemuñir y Matías Seda

**Usuario del equipo en CodaLab:**

**Fecha límite de entrega 📆:** Jueves 14 de Abril.

**Tiempo estimado de dedicación:**

## **Objetivo**

En esta tarea grupal participarán de una competencia estilo [Kaggle](https://www.kaggle.com/) pero utilizando la página [CodaLab](https://codalab.org/). El objetivo es trabajar en la clasificación de tweets  según su intensidad de emoción, esto corresponde a la task de clasificación de texto. 

Tendrán a su disposición 4 datasets de tweets con distintas emociones: `anger`, `fear`, `sadness` y `joy`. Deberán crear un clasificador para cada uno de estos datasets que indique la intensidad de dicha emoción en sus tweets (`low`, `medium`, `high`). 

## **Instrucciones**



- La competencia consiste en resolver 4 problemas de clasificación **distintos**, cada uno con tres clases posibles. Por cada problema deberán crear un clasificador distinto. La evaluación de la competencia se realiza en base a 3 métricas: `AUC`, `Kappa` y `Accuracy`. Los mejores puntajes en cada una de estas métricas serán quienes ganen la competencia.

- Para comenzar se les entregará en este notebook la estructura del informe a entregar y el código de un baseline con el cuál pueden comparar los resultados de sus experimentos. Este baseline consiste en la creación de features y una clasificación simple siguiendo el paradigma de Machine Learning empírico. Los puntajes obtenidos por el baseline los pueden visualizar en el link de CodaLab buscando al usuario **cc6205**. Esperamos que los superen fácilmente 😉

- Para participar, deben registrarse en CodaLab y luego ingresar a la competencia usando el siguiente [link](https://competitions.codalab.org/competitions/30330?secret_key=b2ed0be7-bf23-42a1-9400-f85fa1b7bae7). 

- Está permitido hacer grupos de máximo 4 alumnos. Cada grupo debe tener un nombre de equipo, para ello en CodaLab pueden dirigirse a settings y luego cambiar el Team Name. Sólo una persona debe administrar la cuenta del grupo y se verificará que no se hayan creado múltiples cuentas por grupo.

- En total pueden hacer un **máximo de 4 submissions**, hagan muchos experimentos probando en el conjunto de test antes de realizar el envío.

- Es importante que hagan varios experimentos incorporando técnicas como [cross-validation](https://es.wikipedia.org/wiki/Validaci%C3%B3n_cruzada#:~:text=La%20validaci%C3%B3n%20cruzada%20o%20cross,datos%20de%20entrenamiento%20y%20prueba.) o [random sampling](https://towardsdatascience.com/the-5-sampling-algorithms-every-data-scientist-need-to-know-43c7bc11d17c) antes de enviar sus predicciones a CodaLab, ya que les puede dar un mejor indicio del nivel de generalización de sus modelos. 

- Asegúrense que la distribución de las clases sea balanceada en las particiones de training y testing debido a que existe un desbalanceo. 

- Verificar que el formato de la submission coincida con el de la competencia. De lo contrario, se les será evaluado incorrectamente ya que el Script de evaluación espera como input dicho formato. En el código de las métricas pueden verificar cómo son los inputs.

- No se limiten a los contenidos vistos ni a scikit ni a este baseline. No tienen restricciones entre utilizar Deep Learning o Machine Learning empírico. Si reutilizan gran cantidad de código de alguna página por favor mostrar la referencia en su código. ¡Usen todo su conocimiento e ingenio en mejorar sus sistemas para poder ganar la competencia! 

- **Es requisito entregar el reporte con el código y haber participado en la competencia para ser evaluado. Un código sin reporte o un reporte sin código serán evaluados con la nota mínima.**

- Todas las dudas escríbanlas en el canal de Discord de competencias. Los emails que lleguen al equipo docente serán remitidos a ese medio.

**Importante**: Recuerden poner su nombre y el de su usuario o de equipo (en caso de que aplique) en el reporte. NO serán evaluados Notebooks sin nombre.

----------------------------------------

## **Reporte a entregar**

Uno de los puntos claves en la evaluación de esta competencia es elaborar un informe claro y preciso, argumentando las decisiones tomadas al momento de crear sus modelos para que el cuerpo docente pueda entenderlos. La estructura que debe contener es la siguiente:

1.	**Introducción**: Presentar brevemente el problema a resolver, incluyendo la formalización de la task (cómo son los inputs y outputs del problema) y los desafíos que ven al analizar el corpus entregado. (**0.5 puntos**)
2.	**Representaciones**: Describir los atributos y representaciones usadas como entrada de los clasificadores, **recordar** que para entrenar modelos el input debe tener su representación numérica. Si bien, con Bag of Words (**baseline**) ya se comienzan a percibir buenos resultados, pueden mejorar su evaluación agregando más atributos y representaciones diseñadas a mano, sean lo más creativos posible. Más abajo encontrarán una lista de estos posibles atributos que les podrá ser de utilidad. (**1 punto**)
3.	**Algoritmos**: Describir **brevemente** los algoritmos de clasificación usados, tanto si fueron algoritmos ya vistos en clases o bien arquitecturas de Deep Learning. (**0.5 puntos**)
4.	**Métricas de evaluación**: Describir brevemente las métricas utilizadas en la evaluación, indicando qué miden y su interpretación. (**0.5 puntos**)

5.  **Diseño experimental**: Esta es una de las secciones más importantes del reporte. Deben describir minuciosamente los experimentos que realizarán en la siguiente sección. Describir las variables de control que manejarán, algunos ejemplos pueden ser: Los parámetros de los clasificadores, parámetros en las funciones con que procesan los textos y los transforman, parámetros para el cross-validation, particiones de datos utilizadas, etc. En caso que utilicen redes neuronales, ser claros con el conjunto de hiperparámetros que probarán, la decisión en las funciones de optimización, función de pérdida,  regulación, etc. Básicamente explicar qué es lo que veremos en la siguiente sección.
(**1 punto**)

6.	**Experimentos**: Incluyan todo el código de sus experimentos aquí. ¡Es vital haber realizado varios experimentos para sacar una buena nota! (**1.5 puntos**)

7. **Resultados**: Comparar los resultados obtenidos utilizando diferentes algoritmos y representaciones.  Pueden mostrar los resultados sobre la partición de validación en caso que la generen o sobre los resultados del conjunto de testing. Mostrar los resultados en alguna tabla, pueden poner aquí también los resultados obtenidos al realizar la submission. (**0.5 puntos**)

8. **Conclusiones**:  Discutir resultados, proponer trabajo futuro. (**0.5 puntos**)

## **Descripción Baseline**

` `
El baseline de la sección **Experimentos** contiene un código básico que resuelve la task y es el modelo subido por **cc6205-baseline** a la competencia. Pueden modificar el código como quieran siempre y cuando no cambien las funciones de las métricas y el formato en que se suben los archivos a la competencia, ya que ese es el formato en que realizamos la evaluación. En concretro, el baseline hace lo siguiente:
` `

- Obtiene los datasets desde el repositorio del curso.
- Divide los datasets en train (datos que están etiquetados) y target set (datos no etiquetados para la competencia). Además, por cada dataset en train, se divide en un conjunto de entrenamiento y uno de prueba. Aquí la mejor práctica sería cambiar el código para obtener un conjunto de entrenamiento, validación y prueba.

- Crea un Pipeline que: 
    - Crea features personalizados para la representación numérica.
    - Transforma los dataset a bag of words (BoW).  
    - Entrena un clasificador usando cada train set.
- Clasifica y evalua el sistema creado usando el test set.
- Clasifica el target set.
- Genera una submission con el target en formato zip en el directorio en donde se está ejecutando el notebook. 

` `

Algunas pistas sobre como mejorar el rendimiento de los sistemas que creen. (Esto tendrá mas sentido cuando vean el código)

- **Vectorizador**: Investigar los modulos de `nltk`, en particular, `TweetTokenizer`, `mark_negation` para reemplazar los tokenizadores. También, el parámetro `ngram_range` (Ojo que el clf naive bayes no debería usarse con n-gramas, ya que rompe el supuesto de independencia). Además, implementar los atributos que crean útiles desde el listado del enunciado. Investigar también el vectorizador tf-idf.

- **Clasificador**: Investigar otros clasificadores más efectivos que naive bayes. Estos deben poder retornar la probabilidad de pertenecia de las clases (ie: implementar la función `predict_proba`).

- **Features**: Recuerden que pueden implementar todas las features que se les ocurra! Aquí les adjuntamos algunos ejemplos:
    -	Word n-grams.
    -	Character n-grams. 
    -	Part-of-speech tags.
    -	Sentiment Lexicons (Lexicon = A set of words with a label or associated value.).
        - Count the number of positive and negative words within a sentence.
        - If the lexicon has associated intensity of feeling (for example in a decimal), then take the average of the intensity of the sentence according to the feeling, the sum, etc.
        -	A good lexicon of sentiment: [Bing Liu](http://www.cs.uic.edu/~liub/FBS/opinion-lexicon-English.rar) 
        - A reference with a lot of [sentiment lexicons](https://medium.com/@datamonsters/sentiment-analysis-tools-overview-part-1-positive-and-negative-words-databases-ae35431a470c). 
    -	The number of elongated words (words with one character repeated more than two times).
    -	The number of words with all characters in uppercase.
    -	The presence and the number of positive or negative emoticons.
    -	The number of individual negations.
    -	The number of contiguous sequences of dots, question marks and exclamation marks.
    -	Word Embeddings: Here are some good ideas on how to use them.
    https://stats.stackexchange.com/questions/221715/apply-word-embeddings-to-entire-document-to-get-a-feature-vector

- **Reducción de dimensionalidad**: También puede serles de ayuda. Referencias [aquí](https://scikit-learn.org/stable/modules/unsupervised_reduction.html).

- Por último, pueden encontrar mas referencias de cómo mejorar sus features, el vectorizador y el clasificador [aquí](https://affectivetweets.cms.waikato.ac.nz/benchmark/).

# **Entregable.**

## **1. Introducción**

    Escriba su introducción aquí

## **2. Representaciones**

    Escriba sus decisiones para la representación

## **3. Algoritmos**

    Explique los algoritmos utilizados

## **4. Métricas de Evaluación**

- AUC: ...
- Kappa: ...
- Accuracy: ...


## **5. Diseño experimental**

    Descripción de la metodología 

## **6. Experimentos**

### Importar librerías

In [22]:
import pandas as pd
import os
import numpy as np
import shutil
import codecs
import nltk
import re
import gensim

from functools import reduce

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.metrics import confusion_matrix, cohen_kappa_score, classification_report, accuracy_score, roc_auc_score
from sklearn.model_selection import train_test_split
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

nltk.download('opinion_lexicon')
nltk.download('wordnet')
nltk.download('omw-1.4')

from nltk.corpus import opinion_lexicon, stopwords
from nltk import TweetTokenizer, download
from nltk.stem import SnowballStemmer, WordNetLemmatizer
from nltk.sentiment.vader import SentimentIntensityAnalyzer

from nltk.stem import WordNetLemmatizer


[nltk_data] Downloading package opinion_lexicon to
[nltk_data]     /home/tridimensional/nltk_data...
[nltk_data]   Package opinion_lexicon is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     /home/tridimensional/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     /home/tridimensional/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


### Definir métodos de evaluación (**NO tocar este código**)

Estas funciones están a cargo de evaluar los resultados de la tarea. No deberían cambiarlas.


In [3]:
def auc_score(test_set, predicted_set):
    high_predicted = np.array([prediction[2] for prediction in predicted_set])
    medium_predicted = np.array(
        [prediction[1] for prediction in predicted_set])
    low_predicted = np.array([prediction[0] for prediction in predicted_set])
    high_test = np.where(test_set == 'high', 1.0, 0.0)
    medium_test = np.where(test_set == 'medium', 1.0, 0.0)
    low_test = np.where(test_set == 'low', 1.0, 0.0)
    auc_high = roc_auc_score(high_test, high_predicted)
    auc_med = roc_auc_score(medium_test, medium_predicted)
    auc_low = roc_auc_score(low_test, low_predicted)
    auc_w = (low_test.sum() * auc_low + medium_test.sum() * auc_med +
             high_test.sum() * auc_high) / (
                 low_test.sum() + medium_test.sum() + high_test.sum())
    return auc_w


def evaluate(predicted_probabilities, y_test, labels, dataset_name):
    # Importante: al transformar los arreglos de probabilidad a clases,
    # entregar el arreglo de clases aprendido por el clasificador.
    # (que comunmente, es distinto a ['low', 'medium', 'high'])
    predicted_labels = [
        labels[np.argmax(item)] for item in predicted_probabilities
    ]

    print('Confusion Matrix for {}:\n'.format(dataset_name))
    print(
        confusion_matrix(y_test,
                         predicted_labels,
                         labels=['low', 'medium', 'high']))

    print('\nClassification Report:\n')
    print(
        classification_report(y_test,
                              predicted_labels,
                              labels=['low', 'medium', 'high']))
    # Reorder predicted probabilities array.
    labels = labels.tolist()
    
    predicted_probabilities = predicted_probabilities[:, [
        labels.index('low'),
        labels.index('medium'),
        labels.index('high')
    ]]
    
    
    auc = round(auc_score(y_test, predicted_probabilities), 3)
    print("Scores:\n\nAUC: ", auc, end='\t')
    kappa = round(cohen_kappa_score(y_test, predicted_labels), 3)
    print("Kappa:", kappa, end='\t')
    accuracy = round(accuracy_score(y_test, predicted_labels), 3)
    print("Accuracy:", accuracy)
    print('------------------------------------------------------\n')
    return np.array([auc, kappa, accuracy])

### Datos



In [4]:
# Datasets de entrenamiento.
train = {
    'anger': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/train/anger-train.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity']),
    'fear': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/train/fear-train.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity']),
    'joy': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/train/joy-train.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity']),
    'sadness': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/train/sadness-train.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity'])
}
# Datasets que deberán predecir para la competencia.
target = {
    'anger': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/target/anger-target.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity'], na_values=['NONE']),
    'fear': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/target/fear-target.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity'], na_values=['NONE']),
    'joy': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/target/joy-target.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity'], na_values=['NONE']),
    'sadness': pd.read_csv('https://raw.githubusercontent.com/dccuchile/CC6205/master/assignments/assignment_1/data/target/sadness-target.txt', sep='\t', names=['id', 'tweet', 'class', 'sentiment_intensity'], na_values=['NONE'])
}

In [5]:
# Ejemplo de algunas filas aleatorias del dataset etiquetado:
train['anger'].sample(5)

id                                              tweet  class  \
454  10454  @LaraTheIrish &lt; feel everything. She would ...  anger   
837  10837  Follow me on Snapchat 👻 Carlabtst15 #snapchat ...  anger   
77   10077  What the fuck am I supposed to do with no lunc...  anger   
924  10924  Anger, resentment, and hatred are the destroye...  anger   
414  10414  i can't believe they're all messing up my favo...  anger   

    sentiment_intensity  
454              medium  
837                 low  
77                 high  
924              medium  
414              medium

In [6]:
# Ejemplo de algunas filas aleatorias del dataset no etiquetado
target['anger'].sample(5)

id                                              tweet  class  \
639  11580  @Cass_Pennant @frankbrunoboxer @WestHamUtd @da...  anger   
574  11515  the Bushes were good with the forces of resent...  anger   
604  11545  Men in rage strike those that wish them best. ...  anger   
291  11232  @Thatguy_dree @RecklessWonder_ neither one of ...  anger   
575  11516  I'd pay good money to watch someone slap that ...  anger   

     sentiment_intensity  
639                  NaN  
574                  NaN  
604                  NaN  
291                  NaN  
575                  NaN

### Analizar los datos 

En esta sección analizaremos el balance de los datos. Para ello se imprime la cantidad de tweets de cada dataset agrupados por la intensidad de sentimiento.

In [7]:
for dataset_name in train:
    print(f'Dataset: {dataset_name} \n', train[dataset_name].groupby(['sentiment_intensity']).size())
    print('----------------------------------------------------------\n')

Dataset: anger 
 sentiment_intensity
high      163
low       161
medium    617
dtype: int64
----------------------------------------------------------

Dataset: fear 
 sentiment_intensity
high      270
low       288
medium    699
dtype: int64
----------------------------------------------------------

Dataset: joy 
 sentiment_intensity
high      195
low       219
medium    488
dtype: int64
----------------------------------------------------------

Dataset: sadness 
 sentiment_intensity
high      197
low       210
medium    453
dtype: int64
----------------------------------------------------------



### Custom Features 

Para crear features personalizadas implementaremos nuestros propios Transformers (estandar de scikit para crear nuevas features entre otras cosas). Para esto:

1. Creamos nuestra clase Transformer extendiendo BaseEstimator y TransformerMixin. En este ejemplo, definiremos `CharsCountTransformer` que cuenta caracteres relevantes ('!', '?', '#', '@') en los tweets.
2. Definimos una función cómo `get_relevant_chars` que opera por cada tweet y retorna un arreglo.
3. Hacemos un override de la función `transform` en donde iteramos por cada tweet, llamamos a la función que hicimos antes y agregamos sus resultados a un arreglo. Finalmente lo retornamos.

Esto nos facilitará el trabajo mas adelante. Una Guia completa de las transformaciones predefinidas en scikit pueden encontrarla [aquí](https://scikit-learn.org/stable/data_transforms.html).



In [23]:
import codecs
import numpy as np
from abc import ABC, abstractmethod


class CustomTokenizer:
    def __init__(self):
        self.tweet_tokenizer = TweetTokenizer()

    def tokenizer_tweet(self, tweet):
        return self.tweet_tokenizer.tokenize(tweet)
    
    def tokenizer_with_stemming_doc(self, doc):
        stemmer = SnowballStemmer('english')
        doc_tokenize = list(map(lambda x: self.tweet_tokenizer.tokenize(x), doc))
        doc_tokenize = reduce(lambda x,y: x + y, doc_tokenize, [])
        doc_tokenize = reduce(lambda l, el: l + re.split(r"(\w+)", el), doc_tokenize, [])
        tweet = list(filter(lambda x: x != '' and  x.isalpha(), doc_tokenize))
        stop_words = set(stopwords.words('english'))
        tweet = list(filter(lambda x: x not in stop_words, doc_tokenize))
        return [stemmer.stem(word) for word in doc_tokenize]
    
    def tokenizer_with_lemmatization_tweet(self, tweet):
        lemmatizer = WordNetLemmatizer()
        tweet = self.tweet_tokenizer.tokenize(tweet)
        tweet = reduce(lambda l, el: l + re.split(r"(\w+)", el), tweet, [])
        tweet = list(filter(lambda x: x != '', tweet))
        
        return [lemmatizer.lemmatize(word) for word in tweet]


class Feature(ABC):
    def __init__(self):
        self.tokenizer = CustomTokenizer()
        
    @abstractmethod
    def get_feature(self, tweet: list):
        pass

class Sentiments(Feature):
    # override
    def get_feature(self, tweet: str):
        values = list(SentimentIntensityAnalyzer().polarity_scores(tweet).values())
        values = list(map(lambda x: x if x > 0 else 0, values))
        return list(map(lambda x: int(x*10), values))
    
                    
class Lexicon(Feature):
    # override
    def get_feature(self, tweet: str):
        negative_words = set(opinion_lexicon.negative())
        positive_words = set(opinion_lexicon.positive())
        
        negatives = 0
        positives = 0
        
        tweet = self.tokenizer.tokenizer_with_lemmatization_tweet(tweet)
        for word in tweet:
            if word in negative_words:
                negatives += 1
            if word in positive_words:
                positives += 1

        return [positives, negatives]


class ElongatedWords(Feature):
    # override
    def get_feature(self, tweet: str):

        elongated_words = 0
        tweet = self.tokenizer.tokenizer_tweet(tweet)
        for word in tweet:
            current_letter = word[0]
            count = -1
            for letter in word:
                if letter == current_letter:
                    count += 1
                else:
                    if count >= 4:
                        elongated_words += 1
                        current_letter = letter
                        count = 0

        return [elongated_words]

    
class EmojiLexicon(Feature):
    def __init__(self):
        self.df = pd.read_csv(
            "emoji-sentiment-ranking/emoji-sentiment-data.csv",
            index_col='Emoji'
        )
        self.df['sentiment-score'] = (self.df.Positive - self.df.Negative) / self.df.Occurrences
        self.df['neut-score'] = (self.df.Neutral) / self.df.Occurrences
        self.df = self.df.loc[self.df['neut-score'] <= 0.3]
        
    def get_feature(self, tweet: str):
        positives = 0
        negatives = 0
        positive_score = 0
        negative_score = 0
        for word in tweet:
            if word in self.df.index:
                score = self.df.loc[word, 'sentiment-score']
                positives += 1 if score >= 0.01 else 0
                negatives += 1 if score <= -0.01 else 0
                positive_score += score if score >= 0.01 else 0
                negative_score += abs(score) if score <= -0.01 else 0
        return [positives, negatives, positive_score, negative_score]

class CharsCount(Feature):
    # override
    def get_feature(self, tweet: str):
        chars = ['#', '!', '¡', '?', '¿', '@', '.', '.']
        return list(map(lambda x: tweet.count(x), chars))


class UpperCount(Feature):
    # override
    def get_feature(self, tweet: str):
        
        tweet = self.tokenizer.tokenizer_tweet(tweet)
        upper_count = 0
        for word in tweet:
            if word.isupper():
                upper_count += 1
        return [upper_count]


class CustomTransformer(ABC, BaseEstimator, TransformerMixin):
    def transform(self, X, y=None):
        chars = []

        for tweet in X:
            features = []

            for class_ in self.classes:
                feature = class_().get_feature(tweet)
                features += feature

            chars.append(features)

        return np.array(chars)

    def fit(self, X, y=None):
        return self



class AngryTransformer(CustomTransformer):
    def __init__(self):
        self.classes = [
            Lexicon,
            ElongatedWords,
            CharsCount, 
            UpperCount,
            Sentiments,
            EmojiLexicon
        ]

In [9]:
# Veamos que sucede si ejecutamos el transformer
sample = train['anger'].sample(5, random_state = 22).tweet
sample_features = AngryTransformer().transform(sample)

# Se puede verificar que el conteo de símbolos es consistente con el transformer creado.
print(f'Tweet original: {sample.iloc[0]}')
print(f'Features creados: {sample_features[0]}')

Tweet original: @cineworld 'Congratulations your Free 1 month has been activated' Then charges £34.80 the same month. Absolutely furious 😡
Features creados: [0 1 0 0 0 0 0 0 1 2 2 0 2 4 2 2 0 1]


### Definir la representación y el clasificador

Para esto, definiremos Pipelines. Un `Pipeline` es una lista de transformaciones y un estimador(clasificador) ubicado al final el cual define el flujo que seguiran nuestros datos dentro del sistema que creemos. Nos permite ejecutar facilmente el mismo proceso sobre todos los datasets que usemos, simplificando así nuestra programación.

El pipeline más básico que podemos hacer es transformar el dataset a Bag of Words y después usar clasificar el BoW usando NaiveBayes:

```python
    Pipeline([('bow', CountVectorizer()), ('clf', MultinomialNB())])
```


Ahora, si queremos usar nuestra transformación para agregar las features que creamos, usaremos `FeatureUnion`. Esta simplemente concatenará los vectores resultantes de ejecutar BoW y los Transformer en un solo vector.

```python
    Pipeline([('features',FeatureUnion([('bow', CountVectorizer()),
                                        ('chars_count',CharsCountTransformer())])),
              ('clf', MultinomialNB())])

```





Recuerden que cada pipeline representa un sistema de clasificación distinto. Por lo mismo, deben instanciar uno por cada problema que resuelvan. De lo contrario, podrían solapar resultados.  Para esto, les recomendamos crear los pipeline en distintas funciones, como la siguiente:

In [66]:
from sklearn.feature_extraction.text import TfidfVectorizer

def get_experiment_1_pipeline():
        return Pipeline([('features',
                      FeatureUnion([
                          ('tfidf', TfidfVectorizer()
                          ),
                          ('anger_features', AngryTransformer()),
                                    ])), ('clf', RandomForestClassifier())])
    
def get_experiment_2_pipeline():
        return Pipeline([('features',
                      FeatureUnion([
                          ('tfidf', TfidfVectorizer(tokenizer=CustomTokenizer().tokenizer_with_stemming_doc)
                          ),
                          ('anger_features', AngryTransformer()),
                                    ])), ('clf', RandomForestClassifier())])

    
def get_experiment_3_pipeline():
    vectorizer = TfidfVectorizer(ngram_range=(1, 3))
    log_mod = LogisticRegression(solver='lbfgs', multi_class='ovr', max_iter = 10000)   
    return Pipeline([('features', FeatureUnion([('vect', vectorizer), ('features', AngryTransformer())])), 
                      ('clf', log_mod)])


def get_experiment_4_pipeline():
    vectorizer = TfidfVectorizer(ngram_range=(1, 3), tokenizer=CustomTokenizer().tokenizer_with_stemming_doc)
    log_mod = LogisticRegression(solver='lbfgs', multi_class='ovr', max_iter = 10000)   
    return Pipeline([('features', FeatureUnion([('vect', vectorizer), ('features', AngryTransformer())])), 
                      ('clf', log_mod)])


### Ejecutar el pipeline para algún dataset

In [67]:
def run(dataset, dataset_name, pipeline):
    """Creamos el pipeline y luego lo ejecutamos el pipeline sobre un dataset. 
    Retorna el modelo ya entrenado mas sus labels asociadas y los scores obtenidos al evaluarlo."""

    # Dividimos el dataset en train y test, aún no se transforma de Strings a valores numéricos.
    X_train, X_test, y_train, y_test = train_test_split(
        dataset.tweet,
        dataset.sentiment_intensity,
        shuffle=True,
        test_size=0.33)
    
    print(f'# Datos de entrenamiento en dataset {dataset_name}: {len(X_train)}')
    print(f'# Datos de testing en dataset {dataset_name}: {len(X_test)}')

    # Entrenamos el clasificador (Ejecuta el entrenamiento sobre todo el pipeline). 
    # En este caso el Bag of Words es el encargado de transformar de Strings a vectores numéricos.
    pipeline.fit(X_train, y_train)

    # Predecimos las probabilidades de intensidad de cada elemento del set de prueba.
    predicted_probabilities = pipeline.predict_proba(X_test)

    # Obtenemos el orden de las clases aprendidas.
    learned_labels = pipeline.classes_
    
    # Evaluamos:
    scores = evaluate(predicted_probabilities, y_test, learned_labels, dataset_name)
    return pipeline, learned_labels, scores

### Ejecutar el sistema creado por cada train set

Este código crea y entrena los 4 sistemas de clasificación y luego los evalua. Para los experimentos, pueden copiar este código variando el pipeline cuantas veces estimen conveniente.

In [68]:
f_pipelines = [
    get_experiment_1_pipeline,
    get_experiment_2_pipeline,
    get_experiment_3_pipeline,
    get_experiment_4_pipeline
]

classifiers = []
learned_labels_array = []
for i, f_pipeline in enumerate(f_pipelines):
    print(f"Resultados pipeline: {i+1}\n")
    scores_array = []
    # Por cada nombre_dataset, dataset en train ('anger', 'fear', 'joy', 'sadness')
    for dataset_name, dataset in train.items():
        # creamos el pipeline
        pipeline = f_pipeline()
        
        # ejecutamos el pipeline sobre el dataset
        classifier, learned_labels, scores = run(dataset, dataset_name, pipeline)
        # guardamos el clasificador entrenado (en realidad es el pipeline ya entrenado...)
        classifiers.append(classifier)
        
        # guardamos las labels aprendidas por el clasificador
        learned_labels_array.append(learned_labels)
        
        # guardamos los scores obtenidos
        scores_array.append(scores)
    # print avg scores
    print(
        "Average scores:\n\n",
        "Average AUC: {0:.3g}\t Average Kappa: {1:.3g}\t Average Accuracy: {2:.3g}\n\n"
        .format(*np.array(scores_array).mean(axis=0)))

Resultados pipeline: 1

# Datos de entrenamiento en dataset anger: 630
# Datos de testing en dataset anger: 311
Confusion Matrix for anger:

[[  8  33   0]
 [ 13 188  14]
 [  0  52   3]]

Classification Report:

              precision    recall  f1-score   support

         low       0.38      0.20      0.26        41
      medium       0.69      0.87      0.77       215
        high       0.18      0.05      0.08        55

    accuracy                           0.64       311
   macro avg       0.42      0.37      0.37       311
weighted avg       0.56      0.64      0.58       311

Scores:

AUC:  0.645	Kappa: 0.039	Accuracy: 0.64
------------------------------------------------------

# Datos de entrenamiento en dataset fear: 842
# Datos de testing en dataset fear: 415
Confusion Matrix for fear:

[[ 29  69   1]
 [ 11 194  20]
 [  1  58  32]]

Classification Report:

              precision    recall  f1-score   support

         low       0.71      0.29      0.41        99
      me

Confusion Matrix for anger:

[[  4  47   0]
 [  3 192   2]
 [  0  62   1]]

Classification Report:

              precision    recall  f1-score   support

         low       0.57      0.08      0.14        51
      medium       0.64      0.97      0.77       197
        high       0.33      0.02      0.03        63

    accuracy                           0.63       311
   macro avg       0.51      0.36      0.31       311
weighted avg       0.57      0.63      0.52       311

Scores:

AUC:  0.63	Kappa: 0.039	Accuracy: 0.633
------------------------------------------------------

# Datos de entrenamiento en dataset fear: 842
# Datos de testing en dataset fear: 415
Confusion Matrix for fear:

[[ 25  77   0]
 [ 21 200   8]
 [  3  71  10]]

Classification Report:

              precision    recall  f1-score   support

         low       0.51      0.25      0.33       102
      medium       0.57      0.87      0.69       229
        high       0.56      0.12      0.20        84

    accurac

### Predecir los target set y crear la submission

Aquí predecimos los target set usando los clasificadores creados y creamos los archivos de las submissions.

In [69]:
def predict_target(dataset, classifier, labels):
    # Predecir las probabilidades de intensidad de cada elemento del target set.
    predicted = pd.DataFrame(classifier.predict_proba(dataset.tweet), columns=labels)
    
    # Agregar ids
    predicted['id'] = dataset.id.values
    # Reordenar las columnas
    predicted = predicted[['id', 'low', 'medium', 'high']]
    return predicted

In [70]:
predicted_target = {}

# Crear carpeta ./predictions
if (not os.path.exists('./predictions')):
    os.mkdir('./predictions')

else:
    # Eliminar predicciones anteriores:
    shutil.rmtree('./predictions')
    os.mkdir('./predictions')

# por cada target set:
for idx, key in enumerate(target):
    # Predecirlo
    predicted_target[key] = predict_target(target[key], classifiers[idx],
                                           learned_labels_array[idx])
    # Guardar predicciones en archivos separados. 
    predicted_target[key].to_csv('./predictions/{}-pred.txt'.format(key),
                                 sep='\t',
                                 header=False,
                                 index=False)

# Crear archivo zip
a = shutil.make_archive('predictions', 'zip', './predictions')

## **7. Resultados**

    Muestre y analice sus resultados aquí

A continuación una tabla de ejemplo para mostrar los resultados.  En este caso sólo está el experimento del baseline más otro clasificador, pero ustedes debiesen generar una mayor cantidad de experimentos. 

| No. | Approach                       || Dataset   | AUC   | Kappa | Accuracy |
|-----|--------------------------------||-----------|-------|-------|----------|
|     | Features        | Clasifier     |           |       |       |          |
| 0   | bow+chars_count | MultinomialNB | anger     | 0.622 | 0.163 | 0.688    |
|     |                 |               | fear      | 0.597 | 0.091 | 0.559    |
|     |                 |               | joy       | 0.728 | 0.251 | 0.601    |
|     |                 |               | sadness   | 0.645 | 0.166 | 0.581    |
|     |                 |               |**average**| 0.648 | 0.168 | 0.607    |
| 1   | tus features    | tu clasifier  | anger     |       |       |          |
|     |                 |               | fear      |       |       |          |
|     |                 |               | joy       |       |       |          |
|     |                 |               | sadness   |       |       |          |
|     |                 |               |**average**|       |       |          |


## **8. Conclusiones**

    Escriba aquí sus conclusiones